JustSaying.Messaging/MessageProcessingStrategies/Throttled.cs

```diff
@@ -1,109 +1,75 @@
+            _semaphore.Wait();
+            messageProcessingTask.Start();
         }
 
-        public int AvailableWorkers
+        private async Task ReleaseOnCompleted(Func<Task> action)
         {
-            get
+            try
             {
-                int activeTaskCount;
-                lock (_activeTasks)
-                {
-                    activeTaskCount = _activeTasks.Count;
-                }
-                var freeTasks = _maxWorkersProducer() - activeTaskCount;
-                return Math.Max(freeTasks, 0);
+                await action();
+            }
+            finally
+            {
+                _semaphore.Release();
             }
         }
 
+
         public async Task WaitForAvailableWorkers()
         {
-            var watch = System.Diagnostics.Stopwatch.StartNew();
-
-            // wait for some tasks to complete
-            while (AvailableWorkers == 0)
-            {
-                Task[] activeTasksToWaitOn;
-                lock (_activeTasks)
-                {
-                    activeTasksToWaitOn = _activeTasks.Where(x => x != null).ToArray();
-                }
-
-                if (activeTasksToWaitOn.Length == 0)
-                {
-                    continue;
-                }
+            if (_semaphore.CurrentCount != 0)
+                return;
 
-                _messageMonitor.IncrementThrottlingStatistic();
-                await Task.WhenAny(activeTasksToWaitOn).ConfigureAwait(false);
-            }
+            _messageMonitor.IncrementThrottlingStatistic();
 
+            var watch = Stopwatch.StartNew();
+            await AsTask(_semaphore.AvailableWaitHandle);
             watch.Stop();
+
             _messageMonitor.HandleThrottlingTime(watch.ElapsedMilliseconds);
         }
 
-        public void StartWorker(Func<Task> action)
+        private static Task AsTask(WaitHandle waitHandle)
         {
-            // task is named but not yet started
-            var messageProcessingTask = new Task<Task>(action);
-
-            // what happens when it ends
-            messageProcessingTask.Unwrap()
-                .ContinueWith(t => MarkTaskAsCompleted(messageProcessingTask), TaskContinuationOptions.ExecuteSynchronously);
+            var tcs = new TaskCompletionSource<object>();
 
-            // start it
-            MarkTaskAsActive(messageProcessingTask);
-            messageProcessingTask.Start();
-        }
+            ThreadPool.RegisterWaitForSingleObject(
+                waitObject: waitHandle,
+                callBack: (o, timeout) => { tcs.SetResult(null); },
+                state: null,
+                timeout: TimeSpan.FromMilliseconds(int.MaxValue),
+                executeOnlyOnce: true);
 
-        private void MarkTaskAsActive(Task t)
-        {
-            lock (_activeTasks)
-            {
-                _activeTasks.Add(t);
-            }
+            return tcs.Task;

```

Comment:
Yeah, as you say the alternative is to wait (sync or async) then release. This is more semantically correct, but uglier.